scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,622 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the
+// MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2020 Datadog, Inc.
+//
+use crate::channels::ChannelCapacity;
+use futures::Stream;
```

Comment:
I think here and elsewhere, we should stick to lite/core variants of futures, and remove dependency on futures crate eventually

https://docs.rs/futures-core/0.3.6/futures_core/stream/trait.Stream.html
https://docs.rs/futures-lite/1.11.1/futures_lite/stream/index.html

Reply:
All good comments and I'll merge them all, thanks.

About this futures futures-lite, etc. I must admit that although I have liked your patch to do that change earlier and tried to follow suit, I am a bit confused with which is which.

Maybe it is because "futures" sounds so official =)
Both "futures" and "futures-lite" are listed as crates on Cargo and from what you wrote above lite seems to be one of the good ones?

I think the easiest way to solve this once and for all is to just remove the bad one from the list, fixing all the fallout -- as soon as I understand which it ought to be. I'll do that today so don't worry much about it.

Reply:
Yeah, the situation here is confusing. My understanding of the lay of the land:

* `std::future` and `std::task` module provide the absolute bare bones future interface. They essentially just expose compiler hooks required to make the whole construction work. At the moment, they lack certain things which absolutely need to be in std, but are just not designed yet. Specifically, std lacks abstraction for streams and integration of futures and io. Stuff in std is interface types.
* `futures-core` -- these are missing types from std: `Stream`, `AsyncRead`, `AsyncWrite`. This is also an interface crate (it, it's types are your public API)
* `futures` crate re-exports `futures-core`, and provides various convenience methods and combinators. It's an implementation crate (ie, users of your crate are not aware if you are using futures or not, they are aware only of futures-core). `futures` is also a historical crate, the first version of futures predates async/await. As such, it went through a number of redesigns, and each redesign was compromised by the requirement of maintaining backwards compatibility. futures accumulated a lot of cruft.
* `futures-lite` plays exactly the same role as `futures`, but it has a benefit of starting from clean slate, which, as claimed in the readme, are substantional: https://github.com/stjepang/futures-lite.

My takeaways here:

* we definitely should not use *both* `futures` and `futures-lite`
* which one we use is not overly important, as they are impl details (and both bottom out to std/future-core for interface types)
* `futures-lite` seems to be a more compelling alternative. Caveat -- I am not a heavy futures user, so I can't speak from direct experience. However, stjepang is one of the best library designers I know, and I am comfortable just trusting their opinion without much thought :-)

Reply:
futures-lite it is.
A lot of scipio itself was inspired by his work so I am totally cool with that.

So we don't need to list futures-core as a crate right ?(right now it is not).
I'll just go ahead later today and delete futures from the cargo list.

Reply:
I played with this a bit and there is one major showstopper

Correct me if I am missing something, but compare those two:
https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.for_each
https://docs.rs/futures-lite/1.11.1/futures_lite/stream/trait.StreamExt.html#method.for_each

The futures-lite version of `for_each` cannot take an async closure? That means I essentially cannot chain futures in parallel.

It also seems to lack an implementation of `for_each_concurrent` and `try_for_each_concurrent` which I consider essential.

Reply:
This is a bit sad. futures-lite claims to be compatible with futures, but I tried to standardize into one of them - and tried in both directions, and none of the efforts were trivial.

The use of futures-lite seem to be coming mostly from the code we inherited originally from async-io, which makes sense as they are all from the same author.

Most of the things that I have added, unaware of the distinction between them, seemed to ended up being `futures` which hints that maybe I have an unconscious preference towards it.

The lack of `for_each_concurrent` and the fact that `for_each` don't seem to take async functions itself seem like a bit of a deal breaker to me as I've been using those two things heavily (again, unless I am missing something).

Eager to hear your thoughts in the morning but from what I could gather so far the most likely direction would be for us to standardize on `futures`

A lot of the `futures-lite` code is in networking.rs which I'd like to rewrite to use io_uring native anyway so I wouldn't necessarily bother converting at the moment

Reply:
I think `for_each` case is a genuine API cleanup, as original for_each breaks orthogonality. The `then` function can be used for async processing:

```rust
stream.for_each(|x| async { stuff })
==>
stream.then(|x| async { stuff }).for_each(|x| ())
stream.then(|x| async { stuff }).collect::<()>()
```

`for_each_concurrent` is indeed missing: https://github.com/stjepang/futures-lite/issues/26. I *think* this is understandable, as API is a bit weird -- the `limit` parameter looks like it is from another layer of the abstraction. It doesn't look like a primitive combinator. 

It also shouldn't be hard to roll-our-own `for_each_concurrent` as a free-standing function.

I slightly lean towards using futures-lite and rolling our own combinators which don't exist there yet, but this is a weak opinion.

Reply:
fair comment on the usage of `then`.
It is a weird name because it feels like a `map`, and when I was reading futures-lite code using `then` I somehow actually understood that it would apply that to the end result of the stream, fold-like. Some things make a bit more sense now.

But correct me if I am wrong: let's say our channels now implement `futures_lite::Stream` instead of `futures::Stream`
Doesn't that mean that users of the channel now have to use `futures_lite::StreamExt` instead of `futures::StreamExt`?

It is one thing for us to implement our own version of the concurrent functions, I could be persuaded of that. Another is to force that into users.

I actually kind of like `for_each_concurrent` (and it's try counterpart). I am biased because it reminds me a lot of Seastar's `parallel_for_each` which was by far seastar's most useful primitive, except that it had limit control which is something we wanted to add for a long time.

In the vast majority of cases futures are only useful when you are spamming a bunch of them in parallel. Say for instance you are scanning a directory and reading file headers: you want to do it with something like `try_for_each_concurrent`, not `try_for_each` to generate enough parallelism. The limit helps in making sure you don't blow up your memory (although in all fairness, an excellent API would express that in terms of resources and not arbitrary units)

Reply:
> But correct me if I am wrong: let's say our channels now implement futures_lite::Stream instead of futures::Stream
Doesn't that mean that users of the channel now have to use futures_lite::StreamExt instead of futures::StreamExt?

So here's the trick: both `futures_lite::Stream` and `futures::Stream` are re-exports of `futures_core::Stream`:

* https://github.com/rust-lang/futures-rs/blob/67566c65f5bba15dfad9d1d77e16d919bc18875f/futures/src/lib.rs#L441
* https://github.com/stjepang/futures-lite/blob/39affa66a7170ea425861cdab9043b3b2ed17d87/src/stream.rs#L22

They are exactly the same trait! The `Ext` versions are different between the crates, and each `Stream` implements both `Ext`s. So futures vs futures-lite is purely an implementation decsision, the public API is futures_core. 

>I actually kind of like for_each_concurrent (and it's try counterpart).

I do as well. What I am unsure about is whether they belong to the futures abstraction layer, or to the framework (scipio). For example, one problem with `for_each_concurrent` is that the limit is per invocation. So, if you have 10 tasks each of which spawn for_each_concurrent with limit of 10, the total limit would be 100. In scipio, we can, for example (by now means I am claiming that this is actually a good idea) provide global limit on the parallelism, a-la make jobserver, so that *all* concurrent `for_each_concurrent`s execute at most 10 things.

Reply:
fun discussion!!
I had issues converting the channel code in the examples to futures-lite. It seemed that `for_each` required an implementation of Iterator, not Stream but it was late. I will go back to it soon with this newfound knowledge of `then` and see if I can make that work.

If the `Stream` type is the same, I am more open to using -lite everywhere since the application could keep using `for_each_concurrent`. If you grep, you will notice that scipio itself doesn't use it, which makes sense as it is mostly providing primitives. But the application we are building with scipio uses that everywhere...

Going back to the fun part I do think that a limit belongs in the invocation. You can always add a global limit on top of that.

Before I found out about `try_for_each` I was emulating that with Scipio's Semaphore. It wouldn't be a bad idea to come up with a higher level API that processes a stream in parallel and gets passed a `ResourceManager` kind of object. 

A `ResourceManager` can be built on top of a `Semaphore` or multi-tiered `Semaphores` and could be generic enough to support many such use cases.

For instance, let's say I can only allow my file spawning to support 100 parallel files (because say, that's my fd limit) but at the same time only at most 2MB of memory (because that's how much buffer space I allocated for this)

Every file being processed would bump the `ResourceManager` accordingly as they use resources.